jetty-http2/http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java

```diff
@@ -1647,4 +1589,149 @@ private void complete()
+     * claim the slot in the queue. Concurrent threads will only
+     *
+     */
+    private class StreamCreator
+    {
+        private final Queue<Slot> slots = new ArrayDeque<>();
+        private Thread flushing;
+
+        private int priority(PriorityFrame frame, Callback callback)
+        {
+            Slot slot = new Slot();
+            int streamId = frame.getStreamId();
+            if (streamId <= 0)
+            {
+                synchronized (this)
+                {
+                    streamId = localStreamIds.getAndAdd(2);
+                    slots.offer(slot);
+                }
+                frame = new PriorityFrame(streamId, frame);
+            }
+            else
+            {
+                synchronized (this)
+                {
+                    slots.offer(slot);
+                }
+            }
+
+            slot.entry = new ControlEntry(frame, null, callback);
+            flush();
+            return streamId;
+        }
+
+        private void newStream(HeadersFrame frame, Promise<Stream> promise, Stream.Listener listener)
+        {
+            Slot slot = new Slot();
+            try
+            {
+                int streamId = frame.getStreamId();
+                if (streamId <= 0)
+                {
+                    synchronized (this)
+                    {
+                        streamId = localStreamIds.getAndAdd(2);
+                        slots.offer(slot);
+                    }
+                    PriorityFrame priority = frame.getPriority();
+                    priority = priority == null ? null : new PriorityFrame(streamId, priority);
+                    frame = new HeadersFrame(streamId, priority, frame);
+                }
+                else
+                {
+                    synchronized (this)
+                    {
+                        slots.offer(slot);
+                    }
+                }
+
+                fillSlotWithNewLocalStream(slot, frame, promise, listener, streamId);
+            }
+            catch (Throwable x)
+            {
+                removeSlotAndFailPromise(slot, promise, x);
+            }
+        }
+
+        private void push(PushPromiseFrame frame, Promise<Stream> promise, Stream.Listener listener)
+        {
+            Slot slot = new Slot();
+            try
+            {
+                int streamId;
+                synchronized (this)
+                {
+                    streamId = localStreamIds.getAndAdd(2);
+                    slots.offer(slot);
+                }
+                frame = new PushPromiseFrame(streamId, frame);
+
+                fillSlotWithNewLocalStream(slot, frame, promise, listener, streamId);
+            }
+            catch (Throwable x)
+            {
+                removeSlotAndFailPromise(slot, promise, x);
+            }
+        }
+
+        private void fillSlotWithNewLocalStream(Slot slot, Frame frame, Promise<Stream> promise, Stream.Listener listener, int streamId)
```

Comment:
Can we find a better method name? It's not about filling, because it also flushes.
How about just `createLocalStream(...)`?

Reply:
I renamed it to `createLocalStream` despite the fact that this name clashes with the enclosing class' similarly named method. Despite this, I still find that name better.